From b854d9fec6f7e5636ab4742d716c7d848e0ce0d6 Mon Sep 17 00:00:00 2001
From: Remi Gacogne <remi.gacogne@powerdns.com>
Date: Wed, 26 Oct 2016 15:42:27 +0200
Subject: [PATCH] auth: In `Bind2Backend::lookup()`, use the `zoneId` when we
 have it

After the initial lookup corresponding to a `DNSBackend::getAuth()`,
the subsequent ones already have the `zoneId`, so use it instead of
looping on `chopOff()` again. This should be much more efficient.

(cherry picked from commit 937a66255ff05f2e754ef113833e54cc4cf2004b)
---
 modules/bindbackend/bindbackend2.cc | 14 +++++++++++---
 1 file changed, 11 insertions(+), 3 deletions(-)

Index: pdns/modules/bindbackend/bindbackend2.cc
===================================================================
--- pdns.orig/modules/bindbackend/bindbackend2.cc
+++ pdns/modules/bindbackend/bindbackend2.cc
@@ -1028,9 +1028,17 @@ void Bind2Backend::lookup(const QType &q
   bool found=false;
   BB2DomainInfo bbd;
 
-  do {
-    found = safeGetBBDomainInfo(domain, &bbd);
-  } while ((!found || (zoneId != (int)bbd.d_id && zoneId != -1)) && domain.chopOff());
+  if (zoneId != -1) {
+    found = safeGetBBDomainInfo(zoneId, &bbd);
+    if (found) {
+      domain = bbd.d_name;
+    }
+  }
+  else {
+    do {
+      found = safeGetBBDomainInfo(domain, &bbd);
+    } while (!found && domain.chopOff());
+  }
 
   if(!found) {
     if(mustlog)
